packages/jsapi-bootstrap/src/useObjectFetcher.ts

```diff
@@ -0,0 +1,54 @@
+import { VariableDefinition } from '@deephaven/jsapi-types';
+import { useContextOrThrow } from '@deephaven/react-hooks';
+import { createContext } from 'react';
+
+/**
+ * Metadata for an object. Used when needing to fetch an object.
+ */
+export type ObjectMetadata = Record<string, unknown> & VariableDefinition;
+
+export type ObjectFetcher<T = unknown> = (
+  metadata: ObjectMetadata
+) => Promise<T>;
+
+export const ObjectFetcherContext = createContext<ObjectFetcher | null>(null);
+
+/**
+ * Get the metadata for an object definition that is serializable.
+ * @param definition Object definition to get the metadata for
+ * @returns Metadata object that is serializable
+ */
+export function getObjectMetadata(
```

Comment:
Yea it is quite confusing isn't it... so with Core it's pretty straightforward, we only need the variable definition... whereas on Enterprise, we'd need the variable definition and either the query serial, or the session ID. I guess those can just extend the `VariableDefinition` type, e.g.
```
type QueryVariableDefinition = VariableDefinition & { querySerial: string };
type SessionVariableDefinition = VariableDefinition & { sessionId: string };
```
That makes a lot more sense... the complication comes when migrating data and ensuring we don't break existing dashboards though.

Reply:
Okay, so following up on this a bit...
- Decided to use `VariableDescriptor` everywhere, instead of `VariableDefinition`. Convert from `VariableDefinition` to `VariableDescriptor` right away when we have a `VariableDefinition`
  - That ensures we have a spreadable/serializable object for the descriptor
- Intent is for Enterprise to extend that instead of `VariableDefinition` (e.g. `type QueryVariableDescriptor = VariableDescriptor & { querySerial: string}`)
- Only use the `VariableDefinition` if we use it immediately in the `getObject` with the connection we got it from